### ISSUE TITLE
Attribute Setter gets called twice

### ISSUE THREAD

**Issue by Nijusan** [hautelook/AliceBundle#392]
With the change from version one to two i experience another problem while upgrading an existing project. In this project we have an entity with an attribute which should only be set once, so if the setter is called although the attribut is not empty, we throw an exception.

so i migrated the the bundle to version two, migrated the fixtures and ran into this exception. so somehow while loading the fixtures, the setter on this object is called a second time. i narrowed it down to this constellation as an example:

i have two entites, an user and an adress entity:

<details>
<summary>User</summary>

```php
<?php


namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;

/**
 * @package App\Entity
 * @author Dennis Langen <dennis.langen@i22.de>
 *
 * @ORM\Entity()
 */
class User
{
    /**
     * @var integer
     * @ORM\Id()
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var string
     * @ORM\Column(type="string", length=255)
     */
    private $slug;

    /**
     * @var string
     * @ORM\Column(type="string", length=255)
     */
    private $hash;

    /**
     * @return int
     */
    public function getId(): int
    {
        return $this->id;
    }

    /**
     * @param int $id
     */
    public function setId(int $id): void
    {
        $this->id = $id;
    }

    /**
     * @return string
     */
    public function getSlug(): string
    {
        return $this->slug;
    }

    /**
     * @param string $slug
     */
    public function setSlug(string $slug): void
    {
        if (!empty($this->slug)) {
            throw new \LogicException('cant override already initialized slug');
        }
        $this->slug = $slug;
    }

}
```
</details>

<details>
<summary>Address</summary>

```php
<?php


namespace App\Entity;

use Doctrine\ORM\Mapping as ORM;

/**
 * Class Address
 * @package App\Entity
 * @author Dennis Langen <dennis.langen@i22.de>
 */
class Address
{
    /**
     * @var integer
     * @ORM\Id()
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var User
     */
    private $user;

    /**
     * @var string
     * @ORM\Column(type="string", length=255)
     */
    private $street;

    /**
     * Address constructor.
     * @param User $user
     */
    public function __construct(User $user)
    {
        $this->user = $user;
    }

}
```
</details>

I know for the user there is no information for the db, but its just for testing purpose, the interesting fact is the __construct. so for this here are the fixture files:

```yaml
App\Entity\User:

    user_0:
        slug: 'test01'
        hash: 'xyz'
```

```yaml
App\Entity\Address:

    address_0:
        __construct: ["@user_0"]
        street: 'teststraße'
```

and its with this constellation, that the user is referenced here for the construct, that the setSlug of the user is called twice.

is this a known behavior? is it as expected? i'll have a further look into it whats happening, not sure i'll find the answer...

**Comment by Nijusan** [comment]
update: to get more specific on this, its only happening when inserting the user over the __construct. setting the user over a public setter oder with direct on the private attribute with the propertyReflectionAccessor does not run into the same problems

**Comment by theofidry** [comment]
Hi!

Thanks for the report. This may be related to https://github.com/nelmio/alice/issues/851, not 100% sure about this. I suggest you to track that issue and we can test it again here once it's solved.

PS: I did some edits in your report to make it more readable.

**Comment by Nijusan** [comment]
hi! thanks and thanks :) yeah seems to be related, i'll track that issue

**Comment by Nijusan** [comment]
hi @theofidry . i found the problem and it's drilled down to the fact that there is use of __construct in a fixture file.

the process is complex, but u should know :) so i hope u'll understand my explanation. 

normally, without the construct information, the process would somehow with the two entities above:

* DoublePassGenerator calls generateFixtures() for first time, empty classes are generated
* DoublePassGenerator calls generateFixtures() for second time, empty classes gets hydrated with information

but with the construct, the Initiator does it's job and causing problems. Because its still firstPass but the InitatorResolver calls resolveFixtureConstructor() and yeah the spec does have information about a constructor so resolveArguments() is called and with this the FixtureReferenceResolver. In this Resolver the SimpleObjectGenerator is called, we are still in firstPass, but needsCompleteGeneration is set to true, so the hydration is already done for the user.

so when the generateFixtures() is called for the second time (secondPass), then the hydration will fail.
so i think the process is ok, we need to already hydrate the user in the firstPass to get the adress object created in the secondPass. i think what we need is a marker for the user object, that it was already hydrated in the firstPass and does not need to hydrate again.

if its correct, do u have any advice, how u would like to achieve this goal?

**Comment by theofidry** [comment]
>. i think what we need is a marker for the user object, that it was already hydrated in the firstPass and does not need to hydrate again.

Actually this marker already exists: once an object is complete it's a `CompleteObject` or something. But from your explanation I guess there's a check missing somewhere, or maybe the object is not properly "marked" as complete

**Comment by Nijusan** [comment]
yeah ur right. the check exists, but is only called in the CompleteObjectGenerator which calles the SimpleObjectGenerator->generate()...

the resolver calls the simpleObjectGenerator directly, so there is no marking this object as complete. and i am not sure if this would work, because the check if an object is complete checks if its the firstPass... and we are in the first pass, so i think it would still mark it as incomplete.. furthermore the generate in the simpleObject does not check for completed objects?

**Comment by Nijusan** [comment]
so the solution may be to switch the generator in the resolver to the CompleteObjectGenerator instead of the SimpleObjectGenerator? i'll try and let the tests run

**Comment by Nijusan** [comment]
closed due to https://github.com/nelmio/alice/pull/858
